### PR TITLE
Span hooks

### PR DESCRIPTION
**Pull request by pacejackson**
This PR adds "Hooks" into the tracing lifecycle like the [Observers](https://github.com/reddit/baseplate.py/blob/master/baseplate/__init__.py#L32-L86) in `baseplate.py`.  This allows us to hook custom functionality into the Baseplate request lifecycle and the Span lifecyle.  

This also includes hooks for Metrics which allow us to automatically add timers around Spans and success/fail metrics.

### CODE REVIEW

File: tracing/span.go [hunk 1]
```diff
@@ -33,7 +40,7 @@ const (
 type Span struct {
 	tracer *Tracer
 
-	name     string
+	Name     string
```
Comment (fishy) [inline tracing/span.go]: style nit: move the exported fields to the top, before all the unexported fields.

File: tracing/span.go [hunk 2]
```diff
@@ -66,6 +75,22 @@ func newSpan(tracer *Tracer, spanType SpanType) *Span {
...
+	if spanType == SpanTypeServer {
+		onServerSpanCreate(span)
+		span.startSpan()
+	}
+	return span
+}
+
+func (s *Span) startSpan() {
+	for _, hook := range s.hooks {
+		hook.OnStart(s)
```
Comment (fishy) [inline tracing/span.go]: Should we check the error here? If not, do we need the interface to actually return an error?

File: tracing/span.go [hunk 3]
```diff
@@ -175,11 +200,12 @@ func (s *Span) End(ctx context.Context, err error) error {
 	if s.spanType == SpanTypeLocal && s.component != "" {
 		s.SetTag(ZipkinBinaryAnnotationKeyLocalComponent, s.component)
 	}
-
+	for _, hook := range s.hooks {
+		hook.OnEnd(s, err)
```
Comment (fishy) [inline tracing/span.go]: Same question here.

File: tracing/tracer.go [hunk 4]
```diff
@@ -111,9 +111,8 @@ func (t *Tracer) Close() error {
 
 // NewTrace creates a new trace (top level local span).
 func (t Tracer) NewTrace(name string) *Span {
-	span := newSpan(&t, SpanTypeLocal)
-	span.name = name
-	span.traceID = rand.Uint64()
+	span := newSpan(&t, name, SpanTypeLocal)
+	span.NameeID = rand.Uint64()
```
Comment (fishy) [inline tracing/tracer.go]: What is `NameeID`? It's not defined.

File: tracing/span.go (hunk 2 above)
Comment (pacejackson) [reply tracing/span.go]: I was thinking I would use `s.tracer.Logger` to log an error message?

File: tracing/tracer.go (hunk 4 above)
Comment (pacejackson) [reply tracing/tracer.go]: I think this was just a wonky delete, it was supposed to delete one line but it looks like instead I deleted 1/2 of each line.

File: tracing/span.go [hunk 5]
```diff
@@ -66,6 +79,29 @@ func newSpan(tracer *Tracer, spanType SpanType) *Span {
...
+	span := newSpan(tracer, name, SpanTypeServer)
+	onServerSpanCreate(span)
+	span.startSpan()
+	return span
+}
+
+func (s *Span) startSpan() {
+	for _, hook := range s.hooks {
+		if err := hook.OnStart(s); err != nil {
+			log.Error("OnStart hook error: " + err.Error())
```
Comment (fishy) [inline tracing/span.go]: nit: When using the zap logger, prefer to do this instead: `log.Errorw("OnStart hook error", "err", err)`. Or we can use the wrapped simple logger in tracer:

```
if s.tracer.Logger != nil {
  s.tracer.Logger("OnStart hook error: " + err.Error())
}
```

**Pull request merged**